azure/services/groups/client.go

```diff
@@ -63,27 +67,94 @@ func (ac *azureClient) Get(ctx context.Context, name string) (resources.Group, e
+// CreateOrUpdateAsync creates or updates a resource group.
+// Creating a resource group is not a long running operation, so we don't ever return a future.
+func (ac *azureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter) (azureautorest.FutureAPI, error) {
 	ctx, span := tele.Tracer().Start(ctx, "groups.AzureClient.CreateOrUpdate")
 	defer span.End()
 
-	return ac.groups.CreateOrUpdate(ctx, name, group)
+	group, err := ac.resourceGroupParams(ctx, spec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get desired parameters for group %s", spec.ResourceName())
+	} else if group == nil {
+		// nothing to do here
+		return nil, nil
+	}
+
+	_, err = ac.groups.CreateOrUpdate(ctx, spec.ResourceName(), *group)
+	return nil, err
 }
 
-// Delete deletes a resource group. When you delete a resource group, all of its resources are also deleted.
-func (ac *azureClient) Delete(ctx context.Context, name string) error {
-	ctx, span := tele.Tracer().Start(ctx, "groups.AzureClient.Delete")
+// DeleteAsync deletes a resource group asynchronously. DeleteAsync sends a DELETE
+// request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+//
+// NOTE: When you delete a resource group, all of its resources are also deleted.
+func (ac *azureClient) DeleteAsync(ctx context.Context, spec azure.ResourceSpecGetter) (azureautorest.FutureAPI, error) {
+	ctx, span := tele.Tracer().Start(ctx, "groups.AzureClient.DeleteAsync")
 	defer span.End()
 
-	future, err := ac.groups.Delete(ctx, name)
+	future, err := ac.groups.Delete(ctx, spec.ResourceName())
 	if err != nil {
-		return err
+		return nil, err
 	}
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
+	defer cancel()
+
 	err = future.WaitForCompletionRef(ctx, ac.groups.Client)
 	if err != nil {
-		return err
+		// if an error occurs, return the future.
+		// this means the long-running operation didn't finish in the specified timeout.
+		return &future, err
 	}
 	_, err = future.Result(ac.groups)
-	return err
+	// if the operation completed, return a nil future.
+	return nil, err
+}
+
+// IsDone returns true if the long-running operation has completed.
+func (ac *azureClient) IsDone(ctx context.Context, future azureautorest.FutureAPI) (bool, error) {
+	ctx, span := tele.Tracer().Start(ctx, "groups.AzureClient.IsDone")
+	defer span.End()
+
+	done, err := future.DoneWithContext(ctx, ac.groups)
+	if err != nil {
+		return false, errors.Wrap(err, "failed checking if the operation was complete")
+	}
+
+	return done, nil
+}
+
+// resourceGroupParams returns the desired resource group parameters from the given spec.
+func (ac *azureClient) resourceGroupParams(ctx context.Context, spec azure.ResourceSpecGetter) (*resources.Group, error) {
+	var params interface{}
```

Comment:
```suggestion
        ctx, span := tele.Tracer().Start(ctx, "groups.azureClient.resourceGroupParams")
	defer span.End()
	
	var params interface{}
```

Reply:
do we want to be starting spans for functions that don't make any outside calls ? this function is just a getter so execution should be quick.

Reply:
I'd agree, but just below this is `existingRG, err := ac.Get(ctx, spec.ResourceName())`, which does make a call. If you were looking at the spans, you would see the parent span come from the grandparent func. The spans would not tell the code story as well if you don't fill in the span names all the way down. WDYT?

Reply:
oh my bad, I thought this comment was on the paramaters func in spec.go

you're totally right, this deserves a span, adding it

Reply:
added